File: apps/mobile/ios/Runner/StatementDocumentPicker.swift
// THE iOS HALF OF THE DOCUMENT PICKER, AND THE ENTITLEMENT ARGUMENT FOR IT.
//
// This asks the system for ONE DOCUMENT. It does not ask for a folder, for the
// photo library, or for anything the person did not point at.
//
// NO ENTITLEMENT IS ADDED, AND NONE IS REQUIRED. UIDocumentPickerViewController
// hands back a security-scoped URL for the single file the person selected;
// that grant comes from the selection itself, not from a capability the
// application declares. The read-only user-selected-files entitlement people
// reach for here belongs to the macOS App Sandbox and has no iOS counterpart,
// and no Info.plist usage-description key applies either — the picker is not a
// protected data class. This project's Runner target therefore still has no
// entitlements file at all, which is the state a reviewer should find.
//
// WHAT THIS FILE DELIBERATELY NEVER DOES:
//
//   * it never opens a directory. The folder-picking initialiser would return a
//     scope covering files nobody chose;
//   * it takes NO bookmark. `bookmarkData` is not called, so nothing here can
//     re-open the file after this import — the access ends with the read;
//   * it uses OPEN mode, not import-copy (`asCopy: false`). The copy mode
//     writes a duplicate of a person's bank statement into this application's
//     container, and a statement this product left on the device is exactly
//     what the import flow is written not to do;
//   * it balances the security scope exactly. Every successful
//     startAccessingSecurityScopedResource has one matching stop on every exit
//     path, by `defer`;
//   * it returns NO filename and NO path. The Dart port has no field for
//     either. The same rule governs failures: an error carries a code and
//     nothing else, because a platform message routinely contains the path.
//
// DEVICE EXECUTION IS NOT VERIFIED. Nothing in this repository has run this
// code on a device or a simulator. It compiles, its channel contract is
// asserted from Dart against the framework's mock messenger, and its names are
// asserted against the Dart constants by a source-level test. Behaviour against
// a real document provider is unverified and must not be described otherwise.
import Flutter
import UIKit
import UniformTypeIdentifiers

final class StatementDocumentPicker: NSObject {

  /// Must match `documentPickerChannelName` in
  /// lib/features/statement_imports/data/platform_statement_source_picker.dart.
  /// A fixed literal, deliberately not derived from the bundle identifier,
  /// which carries a per-environment suffix.
  static let channelName = "com.kararfinance.app/document_picker"

  /// The one method this channel answers.
  static let methodPick = "pickCsvSource"

  /// The byte bound, supplied by Dart. See `readAtMost`.
  static let argumentMaxBytes = "maxBytes"

  static let keyBytes = "bytes"
  static let keyMediaType = "mediaType"

  static let codeUnavailable = "PICKER_UNAVAILABLE"
  static let codeUnreadable = "SOURCE_UNREADABLE"
  static let codeBusy = "PICKER_BUSY"
  static let codeInvalidRequest = "INVALID_REQUEST"

  /// What a document of unknown declared type is called.
  private static let unknownMediaType = "application/octet-stream"

  /// The types offered to the document provider.
  ///
  /// CONVENIENCE, NOT A CONTROL. A provider may ignore the filter, a file may
  /// be mislabelled, and a person may rename a spreadsheet. The platform
  /// decides what the content actually is by inspecting it and refuses
  /// archives, spreadsheets and binaries on that basis; nothing downstream
  /// trusts a document because the picker offered it.
  ///
  /// `plainText` is the safe fallback: CSV exports are routinely typed as plain
  /// text, and a filter that hid a person's real statement would be a worse
  /// failure than showing one file too many.
  private static let acceptedContentTypes: [UTType] = [.commaSeparatedText, .plainText]

  /// Reads happen here so the interface thread is not held while a file is
  /// pulled out of a provider that may be a network one.
  private let readQueue = DispatchQueue(label: "com.kararfinance.app.document-read")

  private var channel: FlutterMethodChannel?

  /// The call awaiting a chosen document. At most one at a time.
  private var pending: FlutterResult?

  /// The bound the pending call asked for.
  private var pendingMaxBytes = 0

  func register(with messenger: FlutterBinaryMessenger) {
    let channel = FlutterMethodChannel(name: Self.channelName, binaryMessenger: messenger)
    channel.setMethodCallHandler { [weak self] call, result in
      self?.handle(call, result: result)
    }
    self.channel = channel
  }

  private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard call.method == Self.methodPick else {
      result(FlutterMethodNotImplemented)
      return
    }
    if pending != nil {
      // A picker is already open. Nothing was chosen for this request.
      result(FlutterError(code: Self.codeBusy, message: nil, details: nil))
      return
    }
    // The bound is Dart's to state — it mirrors the server's ingestion limit
    // and is asserted there. This side refuses to guess one.
    guard
      let arguments = call.arguments as? [String: Any],
      let requested = arguments[Self.argumentMaxBytes] as? Int,
      requested > 0
    else {
      result(FlutterError(code: Self.codeInvalidRequest, message: nil, details: nil))
      return
    }
    guard let host = Self.topViewController() else {
      // Nothing on screen to present from. A fact about this moment, not a
      // failure of the person's, and no retry of ours can fix it.
      result(FlutterError(code: Self.codeUnavailable, message: nil, details: nil))
      return
    }

    pending = result
    pendingMaxBytes = requested

    // OPEN, NOT IMPORT-COPY: `asCopy: false`. See the file header.
    let picker = UIDocumentPickerViewController(
      forOpeningContentTypes: Self.acceptedContentTypes,
      asCopy: false
    )
    picker.allowsMultipleSelection = false
    picker.delegate = self
    host.present(picker, animated: true)
  }

  /// Answers the pending call, if there still is one, on the platform thread.
  private func answer(_ reply: @escaping (FlutterResult) -> Void) {
    guard let awaiting = pending else { return }
    pending = nil
    reply(awaiting)
  }

  private func finish(with url: URL) {
    let limit = pendingMaxBytes
    readQueue.async { [weak self] in
      let read = Self.readBounded(url: url, maxBytes: limit)
      DispatchQueue.main.async {
        guard let self = self else { return }
        self.answer { reply in
          guard let read = read else {
            // Carries the code and nothing else: a platform message routinely
            // contains the full path, and a path contains a name.
            reply(FlutterError(code: Self.codeUnreadable, message: nil, details: nil))
            return
          }
          reply([
            Self.keyBytes: FlutterStandardTypedData(bytes: read.bytes),
            Self.keyMediaType: read.mediaType,
          ])
        }
      }
    }
  }

  /// Opens the security-scoped URL, reads it under the bound, and releases the
  /// scope. Returns nil when the document could not be read at all.
  private static func readBounded(url: URL, maxBytes: Int) -> (bytes: Data, mediaType: String)? {
    // BALANCED EXACTLY. `startAccessingSecurityScopedResource` may legitimately
    // answer false — a URL this process already has access to does not need the
    // scope — and calling `stop` for a scope that was never started is the bug
    // this flag exists to prevent.
    let scoped = url.startAccessingSecurityScopedResource()
    defer {
      if scoped {
        url.stopAccessingSecurityScopedResource()
      }
    }

    var coordinationError: NSError?
    var outcome: (bytes: Data, mediaType: String)?
    // Coordinated because an open-in-place URL may be owned by another process,
    // and reading it without coordination can observe a half-written file.
    NSFileCoordinator().coordinate(
      readingItemAt: url,
      options: [],
      error: &coordinationError
    ) { readable in
      guard let bytes = readAtMost(readable, maxBytes: maxBytes) else { return }
      outcome = (bytes: bytes, mediaType: declaredMediaType(of: readable))
    }
    if coordinationError != nil {
      return nil
    }
    return outcome
  }

  /// Reads at most `maxBytes` + 1 bytes.
  ///
  /// THE EXTRA BYTE IS THE POINT. The bound is the server's — Dart supplies it
  /// from the constant that mirrors the platform's ingestion policy — and a
  /// file at exactly the bound is acceptable. Reading one byte past it is what
  /// makes an oversized file arrive with a length GREATER than the bound, so
  /// the client refuses it as too large by the rule it already has, instead of
  /// uploading a silently truncated statement as though it were whole.
  ///
  /// The handle is closed on every exit path, by `defer`, rather than left to
  /// a deallocation this code does not control.
  private static func readAtMost(_ url: URL, maxBytes: Int) -> Data? {
    guard let handle = try? FileHandle(forReadingFrom: url) else { return nil }
    defer { try? handle.close() }
    guard let bytes = try? handle.read(upToCount: maxBytes + 1) else { return nil }
    return bytes
  }

  /// What the provider says the file is. Advisory only, never rendered.
  private static func declaredMediaType(of url: URL) -> String {
    guard
      let values = try? url.resourceValues(forKeys: [.contentTypeKey]),
      let type = values.contentType,
      let mime = type.preferredMIMEType
    else {
      return unknownMediaType
    }
    return mime
  }

  /// The view controller a modal can be presented from.
  private static func topViewController() -> UIViewController? {
    let scene = UIApplication.shared.connectedScenes
      .compactMap { $0 as? UIWindowScene }
      .first { $0.activationState == .foregroundActive }
      ?? UIApplication.shared.connectedScenes.compactMap { $0 as? UIWindowScene }.first
    guard var top = scene?.windows.first(where: { $0.isKeyWindow })?.rootViewController else {
      return nil
    }
    while let presented = top.presentedViewController {
      top = presented
    }
    return top
  }
}

extension StatementDocumentPicker: UIDocumentPickerDelegate {

  func documentPicker(
    _ controller: UIDocumentPickerViewController,
    didPickDocumentsAt urls: [URL]
  ) {
    // Single selection is set on the controller, so anything past the first is
    // not something this build asked for and is not read.
    guard let url = urls.first else {
      answer { reply in reply(nil) }
      return
    }
    finish(with: url)
  }

  func documentPickerWasCancelled(_ controller: UIDocumentPickerViewController) {
    // A person changing their mind is not an error, and is never reported as
    // one.
    answer { reply in reply(nil) }
  }
}

File: apps/mobile/android/app/src/main/kotlin/com/kararfinance/app/StatementDocumentPicker.kt
// THE ANDROID HALF OF THE DOCUMENT PICKER, AND THE PERMISSION ARGUMENT FOR IT.
//
// This asks the system for ONE DOCUMENT. It does not ask for storage.
//
// ACTION_OPEN_DOCUMENT is served by the system document provider UI, which
// returns a content URI carrying a read grant for the single file the person
// selected, scoped to this process and revoked when the task ends. It requires
// NO manifest permission whatsoever. That is the whole reason this file exists
// instead of a file-picker dependency: the merged-manifest allow-list in
// test/security/platform_hardening_test.dart names the exact permission set of
// a real build, and a library that contributed a storage or media permission
// to read one chosen file would be the wrong trade.
//
// WHAT THIS FILE DELIBERATELY NEVER DOES:
//
//   * it declares and requests no permission — not the legacy external-storage
//     pair, not the scoped media ones, not the all-files manage permission.
//     There is no requestPermissions call here and none is needed;
//   * it never opens a directory. ACTION_OPEN_DOCUMENT_TREE would grant access
//     to a whole subtree, which is access to files nobody chose;
//   * it takes NO persistable URI grant. FLAG_GRANT_PERSISTABLE_URI_PERMISSION
//     is not set and takePersistableUriPermission is not called, so the grant
//     dies with the task and this application cannot read the file again later;
//   * it copies the file nowhere. The bytes are read into memory, handed to
//     Dart, and dropped. No cache file, no application-support copy;
//   * it returns NO filename and NO path. The Dart port has no field for either
//     and the platform stores neither, so a name could only ever be logged or
//     displayed — and a bank's export is routinely named after the account it
//     belongs to. The same rule governs failures: an error carries a code and
//     nothing else, because a platform message routinely contains the path.
//
// DEVICE EXECUTION IS NOT VERIFIED. Nothing in this repository has run this
// code on a physical device or an emulator. It compiles, its channel contract
// is asserted from Dart against the framework's mock messenger, and its names
// are asserted against the Dart constants by a source-level test. Behaviour
// against a real document provider is unverified and must not be described as
// anything else.
package com.kararfinance.app

import android.content.ContentResolver
import android.content.Intent
import android.net.Uri
import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/**
 * Presents the system document picker and reads the chosen file, once.
 *
 * The host activity owns the lifecycle: it supplies the launcher, forwards the
 * chosen URI, and detaches when the engine goes away.
 */
internal class StatementDocumentPicker : MethodChannel.MethodCallHandler {

    internal companion object {
        /**
         * Must match `documentPickerChannelName` in
         * lib/features/statement_imports/data/platform_statement_source_picker.dart.
         * A fixed literal, deliberately not derived from the application id:
         * the id carries a per-environment suffix and a channel that moved with
         * it would be a configuration bug visible only at runtime.
         */
        const val CHANNEL = "com.kararfinance.app/document_picker"

        /** The one method this channel answers. */
        const val METHOD_PICK = "pickCsvSource"

        /** The byte bound, supplied by Dart. See the note on [readAtMost]. */
        const val ARG_MAX_BYTES = "maxBytes"

        const val KEY_BYTES = "bytes"
        const val KEY_MEDIA_TYPE = "mediaType"

        const val CODE_UNAVAILABLE = "PICKER_UNAVAILABLE"
        const val CODE_UNREADABLE = "SOURCE_UNREADABLE"
        const val CODE_BUSY = "PICKER_BUSY"
        const val CODE_INVALID_REQUEST = "INVALID_REQUEST"

        /**
         * The types offered to the document provider.
         *
         * CONVENIENCE, NOT A CONTROL. A provider may ignore the filter, a file
         * may be mislabelled, and a person may rename a spreadsheet. The
         * platform decides what the content actually is by inspecting it and
         * refuses archives, spreadsheets and binaries on that basis; nothing
         * downstream trusts a document because the picker offered it.
         *
         * `text/plain` is present because CSV exports are routinely served
         * under it, and a filter that hid a person's real statement would be a
         * worse failure than showing one file too many.
         */
        val ACCEPTED_MIME_TYPES = arrayOf(
            "text/csv",
            "text/comma-separated-values",
            "application/csv",
            "text/plain",
        )

        // The intent type: the any-type wildcard. REQUIRED rather than lazy.
        // The platform documents EXTRA_MIME_TYPES as the way to state a set of
        // acceptable types, and documents that the intent's own type must be
        // the wildcard when that extra is used. Narrowing it here instead would
        // silently drop the extra list on the providers that honour it, leaving
        // a filter that is both stricter and less reliable.
        //
        // ASSEMBLED, NOT WRITTEN AS ONE LITERAL. The security suite strips
        // block comments before scanning this tree, and the wildcard written
        // out contains a slash-star: the stripper would take it for the start
        // of a comment and swallow the declarations after it, quietly narrowing
        // a scan that exists to catch what this tree contains.
        const val OPENABLE_TYPE = "*" + "/" + "*"

        /** Read chunk. Bounded so a large file is not staged in one allocation. */
        private const val CHUNK_BYTES = 64 * 1024
    }

    /** How the host activity presents the picker. Returns false if it could not. */
    internal fun interface DocumentLauncher {
        fun launch(intent: Intent): Boolean
    }

    private var channel: MethodChannel? = null
    private var resolver: ContentResolver? = null
    private var launcher: DocumentLauncher? = null

    /** The call awaiting a chosen document. At most one at a time. */
    private var pending: MethodChannel.Result? = null

    /** The bound the pending call asked for. */
    private var pendingMaxBytes: Int = 0

    private var worker: ExecutorService? = null
    private val main = Handler(Looper.getMainLooper())

    fun attach(messenger: BinaryMessenger, resolver: ContentResolver, launcher: DocumentLauncher) {
        detach()
        this.resolver = resolver
        this.launcher = launcher
        this.worker = Executors.newSingleThreadExecutor()
        this.channel = MethodChannel(messenger, CHANNEL).also { it.setMethodCallHandler(this) }
    }

    fun detach() {
        // A pick that was in flight when the engine went away produced no
        // document, so it is answered as a dismissal rather than left to hang.
        // Nothing is reported to the person: they did not choose a file, and
        // nothing went wrong that concerns them.
        pending?.let { result -> runCatching { result.success(null) } }
        pending = null
        channel?.setMethodCallHandler(null)
        channel = null
        resolver = null
        launcher = null
        worker?.shutdown()
        worker = null
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        if (call.method != METHOD_PICK) {
            result.notImplemented()
            return
        }
        if (pending != null) {
            // A picker is already open. Nothing was chosen for this request.
            result.error(CODE_BUSY, null, null)
            return
        }
        // Read as a Number so a bound that arrived as a 64-bit integer is
        // range-checked rather than throwing inside the codec cast.
        val requested = call.argument<Number>(ARG_MAX_BYTES)?.toLong()
        if (requested == null || requested <= 0L || requested >= Int.MAX_VALUE) {
            // The bound is Dart's to state — it mirrors the server's limit and
            // is asserted there. This side refuses to guess one.
            result.error(CODE_INVALID_REQUEST, null, null)
            return
        }
        val maxBytes = requested.toInt()
        val launcher = this.launcher
        if (launcher == null) {
            result.error(CODE_UNAVAILABLE, null, null)
            return
        }

        pending = result
        pendingMaxBytes = maxBytes
        val launched = launcher.launch(openDocumentIntent())
        if (!launched) {
            // No system document provider could be presented on this device.
            pending = null
            result.error(CODE_UNAVAILABLE, null, null)
        }
    }

    /**
     * The intent, written out rather than taken from a contract helper.
     *
     * CATEGORY_OPENABLE is stated explicitly because the port requires it: it
     * restricts the result to documents that can be opened as a stream, which
     * is the only thing this code does with one. The androidx OpenDocument
     * contract does not add it, so using that contract would have quietly
     * dropped a requirement.
     *
     * No FLAG_GRANT_PERSISTABLE_URI_PERMISSION and no ALLOW_MULTIPLE: one
     * document, readable once, for as long as this task lives.
     */
    private fun openDocumentIntent(): Intent =
        Intent(Intent.ACTION_OPEN_DOCUMENT).apply {
            addCategory(Intent.CATEGORY_OPENABLE)
            type = OPENABLE_TYPE
            putExtra(Intent.EXTRA_MIME_TYPES, ACCEPTED_MIME_TYPES)
            putExtra(Intent.EXTRA_ALLOW_MULTIPLE, false)
        }

    /**
     * The host activity's result callback. [uri] is null when the person
     * dismissed the picker.
     */
    fun onDocumentChosen(uri: Uri?) {
        val result = pending ?: return
        val resolver = this.resolver
        val worker = this.worker
        if (uri == null) {
            pending = null
            result.success(null)
            return
        }
        if (resolver == null || worker == null) {
            pending = null
            result.error(CODE_UNREADABLE, null, null)
            return
        }
        val limit = pendingMaxBytes
        // Reading happens off the platform thread; the answer is posted back on
        // it, because a channel result may only be delivered there.
        worker.execute {
            val mediaType = runCatching { resolver.getType(uri) }.getOrNull()
            val bytes = readBounded(resolver, uri, limit)
            main.post {
                val awaiting = pending
                pending = null
                if (awaiting == null) {
                    return@post
                }
                if (bytes == null) {
                    // Carries the code and nothing else: a platform message
                    // routinely contains the full path, and a path contains a
                    // name.
                    awaiting.error(CODE_UNREADABLE, null, null)
                } else {
                    awaiting.success(
                        mapOf(
                            KEY_BYTES to bytes,
                            KEY_MEDIA_TYPE to (mediaType ?: "application/octet-stream"),
                        ),
                    )
                }
            }
        }
    }

    /**
     * Opens the granted URI, reads it under the bound, and closes it.
     *
     * `use` closes the stream on every exit — value, exception or early return —
     * which is what makes the close deterministic rather than left to a
     * collector. Returns null when the document could not be read at all.
     */
    private fun readBounded(resolver: ContentResolver, uri: Uri, maxBytes: Int): ByteArray? =
        try {
            resolver.openInputStream(uri)?.use { stream -> readAtMost(stream, maxBytes) }
        } catch (error: Exception) {
            // Includes a revoked grant, a provider that died mid-read, and a
            // document whose backing file is gone. None of them produced bytes.
            null
        }

    /**
     * Reads at most [maxBytes] + 1 bytes.
     *
     * THE EXTRA BYTE IS THE POINT. The bound is the server's — Dart supplies it
     * from the constant that mirrors the platform's ingestion policy — and a
     * file at exactly the bound is acceptable. Reading one byte past it is what
     * makes an oversized file arrive with a length GREATER than the bound, so
     * the client refuses it as too large by the rule it already has, instead of
     * uploading a silently truncated statement as though it were whole.
     */
    private fun readAtMost(stream: InputStream, maxBytes: Int): ByteArray {
        val limit = maxBytes + 1
        val collected = ByteArrayOutputStream()
        val chunk = ByteArray(CHUNK_BYTES)
        var total = 0
        while (total < limit) {
            val wanted = minOf(chunk.size, limit - total)
            val read = stream.read(chunk, 0, wanted)
            if (read < 0) {
                break
            }
            collected.write(chunk, 0, read)
            total += read
        }
        return collected.toByteArray()
    }
}
